### PR TITLE
Mspro: option to place PlayReady Object <mspr:pro> alongside to <cenc:pssh> in MPEG-DASH's manifest

### PR DESCRIPTION
**Pull request by tim-lansen**
First i've made basic extraction/insertion of PlayReady Object <mspr:pro>.
The second commit is about backwards compatibility, it adds `-allow_mspr_pro` that enables the feature.

### CODE REVIEW

File: packager/mpd/public/mpd_params.h [hunk 1]
```diff
@@ -83,6 +83,10 @@ struct MpdParams {
   /// If enabled, allow switching between different codecs, if they have the
   /// same language, media type (audio, video etc) and container type.
   bool allow_codec_switching = false;
+  /// If enabled, PlayReady Object <mspr:pro> will be inserted into
+  /// <ContentProtection ...> element alongside with <cenc:pssh>
+  /// when using PlayReady protection system
```
Comment (kqyang) [inline packager/mpd/public/mpd_params.h:88]: End with period ".".

File: packager/mpd/base/mpd_utils.cc [hunk 2]
```diff
@@ -417,6 +439,11 @@ void AddContentProtectionElementsHelperTemplated(
     } else {
       drm_content_protection.scheme_id_uri = "urn:uuid:" + entry.uuid();
       if (!entry.pssh().empty()) {
+        if(entry.uuid() == kPlayReadyUUID && protected_content.allow_mspr_pro()) {
+          drm_content_protection.subelements.push_back(
+              GenerateMsprProElement(entry.pssh()));
+          drm_content_protection.value = kContentProtectionValueMSPR20;
+        }
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:446]: Prefer moving it after line 448 so the element is generated after cenc:pssh as in PlayReady spec examples.

File: packager/mpd/base/mpd_utils.cc [hunk 3]
```diff
@@ -354,6 +359,23 @@ Element GenerateCencPsshElement(const std::string& pssh) {
   return cenc_pssh;
 }
 
+// Extract MS PlayReady Object from given PSSH
+// and encode it in base64
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:363]: end it with period "."

File: packager/mpd/base/mpd_utils.cc [hunk 4]
```diff
@@ -354,6 +359,23 @@ Element GenerateCencPsshElement(const std::string& pssh) {
...
+Element GenerateMsprProElement(const std::string& pssh)
+{
+  std::unique_ptr<media::PsshBoxBuilder> b =
+    media::PsshBoxBuilder::ParseFromBox((const uint8_t*)pssh.data(), pssh.size());
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:367]: Can you replace "(const uint8_t*)" with "reinterpret_cast<const uint8_t*>"? We prefer C++ style cast. Same below.

File: packager/mpd/base/mpd_utils.cc [hunk 5]
```diff
@@ -324,6 +325,10 @@ const char kMarlinUUID[] = "5e629af5-38da-4063-8977-97ffbd9902d4";
 // Unofficial FairPlay system id extracted from
 // https://forums.developer.apple.com/thread/6185.
 const char kFairPlayUUID[] = "29701fe4-3cc7-4a34-8c5b-ae90c7439a47";
+// String representation of media::kPlayReadySystemId
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:328]: End it with period "."

File: packager/mpd/base/mpd_utils.cc [hunk 6]
```diff
@@ -324,6 +325,10 @@ const char kMarlinUUID[] = "5e629af5-38da-4063-8977-97ffbd9902d4";
...
+const char kPlayReadyUUID[] = "9a04f079-9840-4286-ab92-e65be0885f95";
+
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:330]: Add a link on the reason?

For example:

// It is RECOMMENDED to include the @value attribute with name and version "MSPR 2.0".
// See https://docs.microsoft.com/en-us/playready/specifications/mpeg-dash-playready#221-general.

File: packager/mpd/public/mpd_params.h [hunk 7]
```diff
@@ -83,6 +83,10 @@ struct MpdParams {
...
+  bool allow_mspr_pro = false;
 };
```
Comment (kqyang) [inline packager/mpd/public/mpd_params.h:91]: Can you add a unit-test in mpd_utils_unittest.cc like ContentProtectionMarlin?

File: packager/mpd/base/mpd_utils.cc [hunk 8]
```diff
@@ -18,6 +18,7 @@
 #include "packager/mpd/base/content_protection_element.h"
 #include "packager/mpd/base/representation.h"
 #include "packager/mpd/base/xml/scoped_xml_ptr.h"
+#include "packager/media/base/protection_system_specific_info.h"
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:21]: Move it after line 16 (to keep it sorted)

File: packager/mpd/base/mpd_utils.cc [hunk 9]
```diff
@@ -354,6 +359,23 @@ Element GenerateCencPsshElement(const std::string& pssh) {
   return cenc_pssh;
 }
 
+// Extract MS PlayReady Object from given PSSH
+// and encode it in base64
+Element GenerateMsprProElement(const std::string& pssh)
+{
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:365]: Can you move it to line 364 to be consistent to other functions?

File: packager/app/mpd_flags.cc [hunk 10]
```diff
@@ -64,3 +64,8 @@ DEFINE_bool(allow_codec_switching,
             "If enabled, allow adaptive switching between different codecs, "
             "if they have the same language, media type (audio, video etc) and "
             "container type.");
+DEFINE_bool(allow_mspr_pro,
```
[the remark this answers is not in the document]
Comment (tim-lansen) [reply packager/app/mpd_flags.cc:67]: Sure

File: packager/app/mpd_flags.cc [hunk 11]
```diff
@@ -64,3 +64,8 @@ DEFINE_bool(allow_codec_switching,
...
+            false,
```
[the remark this answers is not in the document]
Comment (tim-lansen) [reply packager/app/mpd_flags.cc:68]: So the name of the option should be "-suppress_mspr_pro_for_playready"?

File: packager/mpd/base/mpd_utils.cc (hunk 2 above)
Comment (tim-lansen) [reply packager/mpd/base/mpd_utils.cc:446]: Done

File: packager/mpd/base/mpd_utils.cc (hunk 5 above)
Comment (tim-lansen) [reply packager/mpd/base/mpd_utils.cc:328]: Yep

File: packager/mpd/base/mpd_utils.cc (hunk 6 above)
Comment (tim-lansen) [reply packager/mpd/base/mpd_utils.cc:330]: Of course, forgot it, thanks )

File: packager/mpd/base/mpd_utils.cc (hunk 8 above)
Comment (tim-lansen) [reply packager/mpd/base/mpd_utils.cc:21]: I just was thinking what is wrong with includes sections %)
Done.

File: packager/mpd/base/mpd_utils.cc (hunk 9 above)
Comment (tim-lansen) [reply packager/mpd/base/mpd_utils.cc:365]: Sure

File: packager/app/mpd_flags.cc [hunk 12]
```diff
@@ -64,3 +64,8 @@ DEFINE_bool(allow_codec_switching,
...
+            "If enabled, PlayReady Object <mspr:pro> will be inserted into "
+            "<ContentProtection ...> element alongside with <cenc:pssh> "
+            "when using PlayReady protection system");
```
[the remark this answers is not in the document]
Comment (tim-lansen) [reply packager/app/mpd_flags.cc:71]: Of course

File: packager/mpd/public/mpd_params.h (hunk 1 above)
Comment (tim-lansen) [reply packager/mpd/public/mpd_params.h:88]: Ya

File: packager/mpd/public/mpd_params.h (hunk 7 above)
Comment (tim-lansen) [reply packager/mpd/public/mpd_params.h:91]: Sure

File: packager/app/mpd_flags.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (kqyang) [reply packager/app/mpd_flags.cc:68]: I think I prefer include_mspr_pro_for_playready (with default true), but I am fine with suppress_mspr_pro_for_playready (with default false) if you like it better.

[the remark this answers is not in the document]
Comment (tim-lansen) [reply packager/app/mpd_flags.cc:68]: Ok, i chose the shortest way - set default `true` for  `include_mspr_pro_for_playready`

File: packager/mpd/public/mpd_params.h (hunk 7 above)
Comment (tim-lansen) [reply packager/mpd/public/mpd_params.h:91]: Added two unit tests:
ContentProtectionPlayReadyCencMspr
ContentProtectionPlayReadyCenc

**Comment by tim-lansen** [comment]
> 
> 
> @tim-lansen Thanks for working on this feature. I don't see the commit that address the review comments. Are you still working on it? Or is it already done but you forgot to push the commit?

oh sorry i'm not experienced github team member
i often include multiple changes in single commit if they are clear for me...
Sorry for that..

File: packager/mpd/base/mpd_utils.cc [hunk 13]
```diff
@@ -419,6 +441,11 @@ void AddContentProtectionElementsHelperTemplated(
       if (!entry.pssh().empty()) {
         drm_content_protection.subelements.push_back(
             GenerateCencPsshElement(entry.pssh()));
+        if(entry.uuid() == kPlayReadyUUID && protected_content.include_mspr_pro()) {
+          drm_content_protection.subelements.push_back(
+              GenerateMsprProElement(entry.pssh()));
+          drm_content_protection.value = kContentProtectionValueMSPR20;
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:447]: Should we set the value to "MSPR 2.0" regardless of "include_mspr_pro"?

File: packager/mpd/base/mpd_utils.cc (hunk 4 above)
Comment (kqyang) [reply packager/mpd/base/mpd_utils.cc:367]: Not addressed yet.

File: packager/mpd/base/mpd_utils.cc (hunk 3 above)
Comment (kqyang) [reply packager/mpd/base/mpd_utils.cc:364]: Not addressed yet.

File: packager/mpd/base/mpd_utils.cc [hunk 14]
```diff
@@ -354,6 +360,22 @@ Element GenerateCencPsshElement(const std::string& pssh) {
...
 
+// Extract MS PlayReady Object from given PSSH
+// and encode it in base64
+Element GenerateMsprProElement(const std::string& pssh) {
+  std::unique_ptr<media::PsshBoxBuilder> b =
+    media::PsshBoxBuilder::ParseFromBox((const uint8_t*)pssh.data(), pssh.size());
+
+  const std::vector<uint8_t> *p_pssh = &b->pssh_data();
+  std::string base64_encoded_mspr;
+  base::Base64Encode(base::StringPiece((const char*)p_pssh->data(), p_pssh->size()),
```
Comment (kqyang) [inline packager/mpd/base/mpd_utils.cc:371]: Similar to above. Use C++ style cast reinterpret_cast.

File: packager/mpd/base/mpd_utils.cc (hunk 13 above)
Comment (tim-lansen) [reply packager/mpd/base/mpd_utils.cc:447]: I don't think it is necessary.
Also some consumers may expect no 'value'...

File: packager/mpd/base/media_info.proto [hunk 15]
```diff
@@ -95,6 +95,7 @@ message MediaInfo {
     // Specifies the protection scheme: 'cenc', 'cens', 'cbc1', 'cbcs'.
     // "cbca" is also valid which is a place holder for SAMPLE-AES encryption.
     optional string protection_scheme = 3 [default = 'cenc'];
+    optional uint32 include_mspr_pro = 4;
```
[the remark this answers is not in the document]
Comment (tim-lansen) [reply packager/mpd/base/media_info.proto:98]: Done

File: packager/mpd/base/mpd_utils.cc (hunk 14 above)
Comment (tim-lansen) [reply packager/mpd/base/mpd_utils.cc]: Done

**Comment by kqyang** [comment]
Thanks @tim-lansen. The change LGTM. Would you like to update https://github.com/google/shaka-packager/blob/master/AUTHORS and https://github.com/google/shaka-packager/blob/master/CONTRIBUTORS to include your name?

**Comment by tim-lansen** [comment]
> ... Would you like to update https://github.com/google/shaka-packager/blob/master/AUTHORS and https://github.com/google/shaka-packager/blob/master/CONTRIBUTORS to include your name?
Of course )
I will add
`Ivi.ru LLC <*@ivi.ru>` => AUTHORS
and
`Tim Lansen <tim.lansen@gmail.com>` => CONTRIBUTORS
Is it ok that i use my personal e-mail?

File: AUTHORS [hunk 16]
```diff
@@ -26,3 +26,4 @@ Richard Eklycke <richard@eklycke.se>
 Sanil Raut <sr1990003@gmail.com>
 Sergio Ammirata <sergio@ammirata.net>
 The Chromium Authors <*@chromium.org>
+Ivi.ru LLC <*@ivi.ru>
```
Comment (kqyang) [inline AUTHORS:29]: Can you move it to line 20, i.e. keep it sorted in alphabetical order

File: packager/mpd/base/mpd_utils_unittest.cc [hunk 17]
```diff
@@ -116,4 +117,115 @@ TEST_F(MpdUtilsTest, ContentProtectionMarlin) {
...
+        "    <mspr:pro>MDEyMw==</mspr:pro>"
+        "  </ContentProtection>"
+        "  <Representation id='0' bandwidth='0' codecs='avc1' mimeType='video/mp4'/>"
+        "</AdaptationSet>";
+
+    EXPECT_THAT(adaptation_set_.GetXml().get(), XmlNodeEqual(kExpectedOutput));
+}
+
+TEST_F(MpdUtilsTest, ContentProtectionPlayReadyCenc)
+{
```
[the remark this answers is not in the document]
Comment (tim-lansen) [reply packager/mpd/base/mpd_utils_unittest.cc:177]: Yep

File: AUTHORS (hunk 16 above)
Comment (tim-lansen) [reply AUTHORS:29]: Da

**Comment by kqyang** [comment]
@tim-lansen The change has been merged. Thanks for your help in implementing it!

**Comment by tim-lansen** [comment]
@kqyang thank you for patience )

**Pull request merged**